akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/X509Readers.scala

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.remote.artery.tcp.ssl
+
+import java.security.cert.X509Certificate
+
+import akka.annotation.InternalApi
+import javax.naming.ldap.LdapName
+import akka.util.ccompat.JavaConverters._
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+object X509Readers {
+
+  def getAllSubjectNames(cert: X509Certificate): Set[String] = {
+    val maybeCommonName =
+      new LdapName(cert.getSubjectX500Principal.getName).getRdns.asScala.collectFirst {
+        case attr if attr.getType.equalsIgnoreCase("CN") =>
+          attr.getValue.toString
+      }
+
+    val alternates = Option(cert.getSubjectAlternativeNames).map(_.asScala).getOrElse(Nil).collect {
+      // See the javadocs for what this list contains, first element should be an integer,
+      // if that integer is 2, then the second element is a String containing the DNS name.
+      case list if list.get(0) == 2 =>
```

Comment:
can the list be empty? 
`getOrElse(Nil)` above?

Reply:
Good catch. I'm not sure. This was copy pasted. I'll try to create a keyset with a certificate that has an empty list.

Reply:
The return type of `getSubjectAlternativeNames` is a bit convoluted:

- if the certificate has no `SubjectAlternativeNames` (aka `SAN`), then `null` is returned. Then, we'd be running `Nil.collect(...)` so we'd collect over an empty list of `SAN`.
- else, a `Collection<List<?>>` is returned where each item of the `Collection<>` is a `SAN` represented as an heterogeneous list where `list.get(0)` is an integer identifying the SAN type (`DNS`, `EMAIL`, ...) and `list.get(1)` is the value. By contract of `getSubjectAlternativeNames` doing `list.get(0)` and `list.get(1)` is safe.

Reply:
Thanks for clarifying. I missed that it was a collection of a collection